File: Triangulation/simplex.cpp
/**************************************************************
Class: Simplex
Author: Alex Henniges, Tom Williams, Mitch Wilson
Version: June 9, 2008
**************************************************************/

#include "simplex.h" // class's header file
#include <algorithm>

void Simplex::addVertex(int vertex){
  //vector<int>::iterator it = find(localVertices.begin(), localVertices.end(), vertex);
  //if(it == localVertices.end())
    localVertices.push_back(vertex);
}

void Simplex::addEdge(int edge){
  //vector<int>::iterator it = find(localEdges.begin(), localEdges.end(), edge);
  //if(it == localEdges.end())
    localEdges.push_back(edge);
}

void Simplex::addFace(int face){
  //vector<int>::iterator it = find(localFaces.begin(), localFaces.end(), face);
  //if(it == localFaces.end())
    localFaces.push_back(face);
}

void Simplex::addTetra(int tetra){
  //vector<int>::iterator it = find(localTetras.begin(), localTetras.end(), tetra);
  //if(it == localTetras.end())
    localTetras.push_back(tetra);
}

void Simplex::removeVertex(int vertex){ 
  for(int i = 0; i < localVertices.size(); i++){
    if(localVertices[i] == vertex) {
      localVertices.erase(localVertices.begin() + i);
      return;
    }
  }
}

void Simplex::removeEdge(int edge){
  for(int i = 0; i < localEdges.size(); i++){
    if(localEdges[i] == edge) {
      localEdges.erase(localEdges.begin() + i);
      return;
    }
  }
}

void Simplex::removeFace(int face)
{
  for(int i = 0; i < localFaces.size(); i++)
    {
      if(localFaces[i] == face)
	{
	  localFaces.erase(localFaces.begin() + i);
	  return;
	}
    }   
}
void Simplex::removeTetra(int tetra)
{
  for(int i = 0; i < localTetras.size(); i++)
    {
      if(localTetras[i] == tetra)
	{
	  localTetras.erase(localTetras.begin() + i);
	  return;
	}
    }   
}

void Simplex::clearLocals()
{
  localVertices.clear();
  localEdges.clear();
  localFaces.clear();
  localTetras.clear();
}

bool Simplex::isAdjVertex(int vIndex)
{
  for(int i = 0; i < localVertices.size(); i++)
    {
      if(localVertices[i] == vIndex)
	return true;
    }
  return false;
}

bool Simplex::isAdjEdge(int eIndex){
  for(int i = 0; i < localEdges.size(); i++)
    if(localEdges[i] == eIndex)
      return true;
  
  return false;
}

bool Simplex::isAdjFace(int fIndex){
  for(int i = 0; i < localFaces.size(); i++)
    if(localFaces[i] == fIndex)
      return true;
   
  return false;
}

bool Simplex::isAdjTetra(int tIndex){
  for(int i = 0; i < localTetras.size(); i++)
    if(localTetras[i] == tIndex)
      return true;

  return false;
}

vector<int>* Simplex::getLocalVertices(){
  return &localVertices;
}

vector<int>* Simplex::getLocalEdges(){
  return &localEdges;
}

vector<int>* Simplex::getLocalFaces(){
  return &localFaces;
}

vector<int>* Simplex::getLocalTetras(){
  return &localTetras;
}

File: Java/src/Triangulation/Simplex.java
package Triangulation;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public abstract class Simplex {
  protected int serialNumber;
  protected int index;
  private static int simplexCounter = 0;
  protected LinkedList<Vertex> localVertices;
  protected LinkedList<Edge> localEdges;
  protected LinkedList<Face> localFaces;
  protected LinkedList<Tetra> localTetras;
  
  public Simplex(int index) {
    this.index = index;
    serialNumber = simplexCounter;
    simplexCounter++;
    localVertices = new LinkedList<Vertex>();
    localEdges = new LinkedList<Edge>();
    localFaces = new LinkedList<Face>();
    localTetras = new LinkedList<Tetra>();
  }
  
  public int getIndex() {
    return index;
  }
  
  public int getSerialNumber() {
    return serialNumber;
  }
  
  public void setIndex(int index) {
    this.index = index;
  }
  
  // Local simplices retrieval methods
  public List<Vertex> getLocalVertices() {
    return localVertices;
  }
  public List<Edge> getLocalEdges() {
    return localEdges;
  }
  public List<Face> getLocalFaces() {
    return localFaces;
  }
  public List<Tetra> getLocalTetras() {
    return localTetras;
  }
  
  // Adding local simplices
  public void addVertex(Vertex v) {
    localVertices.add(v);
  }
  public void addEdge(Edge e) {
    localEdges.add(e);
  }
  public void addFace(Face f) {
    localFaces.add(f);
  }
  public void addTetra(Tetra t) {
    localTetras.add(t);
  }
  
  
  // Removing local simplices
  public void removeVertex(Vertex v) {
    localVertices.remove(v);
  }
  public void removeEdge(Edge e) {
    localEdges.remove(e);
  }
  public void removeFace(Face f) {
    localFaces.remove(f);
  }
  public void removeTetra(Tetra t) {
    localTetras.remove(t);
  }
  public void clearLocals() {
    localVertices.clear();
    localEdges.clear();
    localFaces.clear();
    localTetras.clear(); 
  }
  
  // Adjacency checks
  public boolean isAdjVertex(Vertex v) {
    return localVertices.contains(v);
  }
  public boolean isAdjEdge(Edge e) {
    return localEdges.contains(e);
  }
  public boolean isAdjFace(Face f) {
    return localFaces.contains(f);
  }
  public boolean isAdjTetra(Tetra t) {
    return localTetras.contains(t);
  }
  
  public boolean equals(Object other) {
    if(other.getClass() != this.getClass()) {
      return false;
    }
    return this.index == ((Simplex) other).index;
  }
  
  public int hashCode() {
    return this.getClass().hashCode() ^ (new Integer(this.index)).hashCode();
  }
  
  public String toString() {
    String type = this.getClass().toString().substring(6);
    if(type.contains(".")) {
      type = type.substring(type.lastIndexOf(".") + 1);
    }
    return type + " " + index;
  }
}
